src/Shared/Components/PrerenderComponentApplicationStore.cs

```diff
@@ -23,33 +25,105 @@ public PrerenderComponentApplicationStore(string existingState)
                 throw new ArgumentNullException(nameof(existingState));
             }
 
-            ExistingState = JsonSerializer.Deserialize<Dictionary<string, byte[]>>(Convert.FromBase64String(existingState)) ??
-                throw new ArgumentException(nameof(existingState));
+            DeserializeState(Convert.FromBase64String(existingState));
+        }
+
+        protected void DeserializeState(byte[] existingState)
+        {
+            var state = JsonSerializer.Deserialize<Dictionary<string, byte[]>>(existingState);
+            if (state == null)
+            {
+                throw new ArgumentException("Could not deserialize state correctly", nameof(existingState));
+            }
+
+            var stateDictionary = new Dictionary<string, ReadOnlySequence<byte>>();
+            foreach (var (key, value) in state)
+            {
+                stateDictionary.Add(key, new ReadOnlySequence<byte>(value));
+            }
+
+            ExistingState = stateDictionary;
         }
 
 #nullable enable
         public string? PersistedState { get; private set; }
 #nullable disable
 
-        public Dictionary<string, byte[]> ExistingState { get; protected set; }
+        public Dictionary<string, ReadOnlySequence<byte>> ExistingState { get; protected set; }
 
-        public Task<IDictionary<string, byte[]>> GetPersistedStateAsync()
+        public Task<IDictionary<string, ReadOnlySequence<byte>>> GetPersistedStateAsync()
         {
-            return Task.FromResult((IDictionary<string, byte[]>)ExistingState);
+            return Task.FromResult((IDictionary<string, ReadOnlySequence<byte>>)ExistingState);
         }
 
-        protected virtual byte[] SerializeState(IReadOnlyDictionary<string, byte[]> state)
+        protected virtual async Task<byte[]> SerializeState(IReadOnlyDictionary<string, ReadOnlySequence<byte>> state)
         {
-            return JsonSerializer.SerializeToUtf8Bytes(state);
+            // System.Text.Json doesn't support serializing ReadonlySequence<byte> so we need to buffer
+            // the data with a memory pool here. We will change our serialization strategy in the future here
+            // so that we can avoid this step.
```

Comment:
We serialize the dictionary directly manually since System.Text.Json doesn't support serializing ReadOnlySequence<byte>.

We do the serialization using the lower level APIs so that we can avoid having to create an extra dictionary and hold on to pooled memory. The way it works is, if the data is already in a single segment (most of the cases) we write it. If it "fits" within the stack, we copy it to a continuous buffer and write it. Otherwise, we rent a buffer from the memorypool, copy it there, write the property and return the buffer to the pool.